File: backend/app/services/capture_service.py
from uuid import UUID

from backend.app.models.capture import Capture, CaptureCreate
from backend.app.repositories.capture_repository import CaptureRepository


class CaptureService:
    """
    Business logic for captures.
    """

    def __init__(self, repository: CaptureRepository):
        self.repository = repository

    def create_capture(self, data: CaptureCreate) -> Capture:
        capture = Capture(
            title=data.title,
            content=data.content,
            capture_type=data.capture_type,
            workspace_id=data.workspace_id,
            session_id=data.session_id,
        )

        return self.repository.create(capture)

    def list_captures(self) -> list[Capture]:
        return self.repository.list_all()

    def get_capture(self, capture_id: UUID) -> Capture | None:
        return self.repository.get_by_id(capture_id)

    def delete_capture(self, capture_id: UUID) -> bool:
        return self.repository.delete(capture_id)

    def assign_workspace(
        self,
        capture_id: UUID,
        workspace_id: UUID | None,
    ) -> Capture | None:
        return self.repository.assign_workspace(
            capture_id=capture_id,
            workspace_id=workspace_id,
        )

File: frontend/lib/services/capture_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/capture.dart';
import 'api.dart';

class CaptureService {
  Future<List<Capture>> getCaptures() async {
    final response = await http.get(
      Uri.parse('${Api.baseUrl}/captures'),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to load captures: ${response.statusCode}',
      );
    }

    final List<dynamic> data = jsonDecode(response.body);

    return data
        .map(
          (item) => Capture.fromJson(
            item as Map<String, dynamic>,
          ),
        )
        .toList();
  }

  Future<Capture> createCapture({
    required String title,
    required String content,
    String? workspaceId,
    String? sessionId,
  }) async {
    final response = await http.post(
      Uri.parse('${Api.baseUrl}/captures'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'title': title,
        'content': content,
        'capture_type': 'text',
        'workspace_id': workspaceId,
        'session_id': sessionId,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to create capture: ${response.statusCode}',
      );
    }

    return Capture.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<void> deleteCapture(String captureId) async {
    final response = await http.delete(
      Uri.parse('${Api.baseUrl}/captures/$captureId'),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to delete capture: ${response.statusCode}',
      );
    }
  }

  Future<Capture> assignWorkspace({
    required String captureId,
    required String? workspaceId,
  }) async {
    final response = await http.patch(
      Uri.parse(
        '${Api.baseUrl}/captures/$captureId/workspace',
      ),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'workspace_id': workspaceId,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to assign workspace: ${response.statusCode}',
      );
    }

    return Capture.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }
}